keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/diff/DualNumber.java

```diff
@@ -350,4 +351,84 @@ private DoubleTensor pluckFromPartial(DoubleTensor from, int... indices) {
             .reshape(pluckShape);
     }
 
+    public DualNumber ifThenElse(BooleanTensor predicate, DualNumber els) {
+        if (predicate.allTrue()) {
+            return new DualNumber(this.value, this.getPartialDerivatives());
+        } else if (predicate.allFalse()) {
+            return new DualNumber(els.value, els.getPartialDerivatives());
+        } else {
+            return pluckFromMixedPredicate(predicate, this, els);
```

Comment:
This slicing and dicing partial derivatives seems like the responsibility of the PartialDerivatives class. There's a tensor trick called broadcasting that can getting you around the need to concat and plucking. This broadcast adds 1 to a zero tensor of the same shape as the partials. This gets you an appropriately shaped mask for the partials. 

`PartialDerivatives.ifThenElse(...)`:
```
    public static PartialDerivatives ifThenElse(BooleanTensor predicate, PartialDerivatives thn, PartialDerivatives els) {
        DoubleTensor trueMask = predicate.toDoubleMask();
        DoubleTensor falseMask = predicate.not().toDoubleMask();

        Map<Long, DoubleTensor> thenPartials = thn.derivativeWithRespectTo;
        Map<Long, DoubleTensor> elsePartials = els.derivativeWithRespectTo;
        Set<Long> wrtUnion = new HashSet<>();
        wrtUnion.addAll(thenPartials.keySet());
        wrtUnion.addAll(elsePartials.keySet());

        Map<Long, DoubleTensor> mixedPartials = new HashMap<>();
        for (Long wrt : wrtUnion) {
            DoubleTensor thnPartial = thenPartials.get(wrt);
            DoubleTensor elsPartial = elsePartials.get(wrt);
            DoubleTensor broadcastedTrueMask;
            DoubleTensor broadcastedFalseMask;

            DoubleTensor newPartial;
            if (thnPartial == null) {
                broadcastedFalseMask = DoubleTensor.zeros(elsPartial.getShape()).plusInPlace(falseMask);
                newPartial = broadcastedFalseMask.timesInPlace(elsPartial);
            } else if (elsPartial == null) {
                broadcastedTrueMask = DoubleTensor.zeros(thnPartial.getShape()).plusInPlace(trueMask);
                newPartial = broadcastedTrueMask.timesInPlace(thnPartial);
            } else {
                broadcastedFalseMask = DoubleTensor.zeros(thnPartial.getShape()).plusInPlace(falseMask);
                broadcastedTrueMask = DoubleTensor.zeros(thnPartial.getShape()).plusInPlace(trueMask);

                newPartial = broadcastedTrueMask.timesInPlace(thnPartial)
                    .plusInPlace(broadcastedFalseMask.timesInPlace(elsPartial));
            }

            mixedPartials.put(wrt, newPartial);
        }

        return new PartialDerivatives(mixedPartials);
    }
```

Reply:
Awesome! I've replaced the concat's and pluck's with this